src/main/java/com/khutircraftubackend/category/CategoryService.java

```diff
@@ -22,95 +21,112 @@
-        }
-
-
-        if (!childCategories.isEmpty() && forceDelete) {
-            for (CategoryEntity child : childCategories) {
-                deleteCategory(child.getId(), forceDelete);
-            }
-            categoryRepository.deleteById(id);
-        } else if(!childCategories.isEmpty()){
-            throw new CategoryDeletionException(CategoryExceptionMessages.CATEGORY_HAS_SUBCATEGORIES_OR_PRODUCTS);
-        }
-    }
+	
+	private final CategoryRepository categoryRepository;
+	private final CategoryMapper categoryMapper;
+	private final FileConverterService fileConverterService;
+	
+	private String handleIcon(MultipartFile iconFile) throws IOException {
+		
+		if (iconFile == null) {
+			return "";
+		}
+		
+		return fileConverterService.convert(iconFile);
+	}
+	
+	private CategoryEntity findCategoryById(Long id) {
+		
+		return categoryRepository.findById(id).orElseThrow(() ->
+				new CategoryNotFoundException(CategoryExceptionMessages.CATEGORY_NOT_FOUND));
+	}
+	
+	public CategoryEntity getCategoryById(Long id) {
+		
+		return findCategoryById(id);
+	}
+	
+	public List<CategoryResponse> getAllRootCategories() {
+		
+		return categoryRepository.findAllByParentCategoryIsNull().stream().map(
+				categoryMapper::toCategoryResponse
+		).collect(Collectors.toList());
+	}
+	
+	public List<CategoryResponse> getAllByParentCategoryId(Long id) {
+		
+		return categoryRepository.findAllByParentCategory_Id(id).stream().map(
+				categoryMapper::toCategoryResponse
+		).collect(Collectors.toList());
+	}
+	
+	@Transactional
+	public CategoryEntity createCategory(CategoryCreateRequest request, MultipartFile iconFile) throws IOException {
+		
+		CategoryEntity category = categoryMapper.toCategoryEntity(request);
+		category.setIconUrl(handleIcon(iconFile));
+		
+		if (request.parentCategoryId() != null) {
+			CategoryEntity parentCategory = findCategoryById(request.parentCategoryId());
+			
+			category.setParentCategory(parentCategory);
+		} else {
+			category.setParentCategory(null);
+		}
+		
+		return categoryRepository.save(category);
+	}
+	
+	
+	@Transactional
+	public CategoryEntity updateCategory(Long id, CategoryUpdateRequest request, MultipartFile iconFile) throws IOException {
+		CategoryEntity existingCategory = findCategoryById(id);
+		
+		if (iconFile != null && !iconFile.isEmpty()) {
+			existingCategory.setIconUrl(handleIcon(iconFile));
+			
+		} else if (request.iconFile() != null) {
+			existingCategory.setIconUrl(handleIcon(iconFile));
+		}
+		
+		categoryMapper.updateCategoryEntity(existingCategory, request);
+		
+		if (request.parentCategoryId() != null) {
+			CategoryEntity parentCategory = findCategoryById(request.parentCategoryId());
+			existingCategory.setParentCategory(parentCategory);
+			
+		} else {
+			
+			existingCategory.setParentCategory(null);
+		}
+		
+		return categoryRepository.save(existingCategory);
+	}
+	
+	public void deleteCategory(Long id) {
+		deleteCategory(id, false);
+	}
```

Comment:
looks like we ended up with method not being used really. 
it's only covered with tests. Controller doesn't need it.
so can be deleted.

Reply:
Deleted.